src/hooks/useDelayedState.ts

```diff
@@ -0,0 +1,30 @@
+import {useRef} from 'react'
+import {useState, useCallback, SetStateAction} from 'react'
+
+/**
+ * @beta
+ */
+export function useDelayedState<S>(
+  initialState: S | (() => S),
+): [S, (nextState: SetStateAction<S>, delay?: number) => void] {
+  const [state, setState] = useState(initialState)
+  const delayedAction = useRef<NodeJS.Timeout | undefined>()
+
+  const onStateChange = useCallback((nextState: SetStateAction<S>, delay?: number) => {
+    const action = () => {
+      setState((prev) => {
+        if (typeof nextState === 'function') {
+          const callback = nextState as (prevState: S) => S
+          return callback(prev)
+        }
+        return nextState
+      })
```

Comment:
Wondering if this can be shortened:

```suggestion
      setState(nextState)
```